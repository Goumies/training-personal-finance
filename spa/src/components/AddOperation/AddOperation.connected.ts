import { Updater, State, connect } from "../../store";
import { AddOperationComponent, UpdateProps, OwnProps } from "./AddOperation";
import { Operation } from "../../interfaces/Operation";

function computeUpdateProps(updateState: (updater: Updater) => void): UpdateProps {
  const onNewOperation = (operation: Operation) => {
    const updater: Updater = (state: State) => {
      return {
        ...state,
        operations: [...state.operations, operation]
      };
    };
    updateState(updater);
  };
  return { onNewOperation };
}

export const AddOperation = connect<OwnProps, {}, UpdateProps>(
  AddOperationComponent,
  { computeUpdateProps }
);
