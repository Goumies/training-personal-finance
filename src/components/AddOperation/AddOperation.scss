.add-operation {
  margin-bottom: 2em;
  width: 300px;
  display: grid;
  grid-gap: 1em;
  justify-content: left;

  .input-amount {
    grid-column: 1 / 2;

    width: 100px;
    border: 1px solid lightgray;
    padding-right: 0.5em;
    text-align: right;
    font-size: 1em;

    &::-webkit-inner-spin-button,
    &::-webkit-outer-spin-button {
      -webkit-appearance: none;
      margin: 0;
    }
  }

  .add-amount {
    grid-column: 2 / 3;
  }
}
