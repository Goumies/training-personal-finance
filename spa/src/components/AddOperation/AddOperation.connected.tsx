import React from "react";
import { StoreContext, Store, Updater, State } from "../../store";
import { AddOperationComponent, OwnProps } from "./AddOperation";
import { Operation } from "../../interfaces/Operation";

export const AddOperation: React.StatelessComponent<OwnProps> = () => (
  <StoreContext.Consumer>
    {(store: Store | undefined) => {
      if (store) {
        const onNewOperation = (operation: Operation) => {
          const updater: Updater = (state: State) => {
            return {
              ...state,
              operations: [...state.operations, operation]
            };
          };
          store.updateState(updater);
        };
        return <AddOperationComponent onNewOperation={onNewOperation} />;
      }
    }}
  </StoreContext.Consumer>
);
