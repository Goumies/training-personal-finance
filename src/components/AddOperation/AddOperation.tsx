import React from "react";
import { Operation as IOperation } from "../../interfaces/Operation";
import uuid from "uuid/v4";
import "./AddOperation.scss";

interface Props {
  onNewOperation(operation: IOperation): void;
}

interface State {
  value: number | undefined;
}

export class AddOperation extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      value: undefined
    };
  }

  private setValue = (value: number | undefined): void => {
    this.setState({
      ...this.state,
      value
    });
  };

  private onInputChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const valueAsString: string = event.target.value;
    if (valueAsString === "") {
      this.setValue(undefined);
    }
    const value: number = parseFloat(valueAsString);
    if (!isNaN(value)) {
      this.setValue(value);
    }
  };

  private onButtonClick = (): void => {
    const amount: number | undefined = this.state.value;
    if (amount !== undefined && amount !== 0) {
      this.props.onNewOperation({
        id: uuid(),
        date: new Date(),
        amount
      });
    }
  };

  render() {
    return (
      <div className="add-operation">
        <input className="input-amount" type="number" onChange={this.onInputChange} />
        <button
          onClick={this.onButtonClick}
          className="add-amount mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent"
        >
          Add
        </button>
      </div>
    );
  }
}
