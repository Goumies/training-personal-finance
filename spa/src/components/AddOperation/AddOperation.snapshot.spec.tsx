import React from "react";
import renderer from "react-test-renderer";
import { AddOperation, Props } from "./AddOperation";

it("renders correctly", () => {
  const props: Props = {
    onNewOperation: jest.fn()
  };
  const tree = renderer.create(<AddOperation {...props} />).toJSON();
  expect(tree).toMatchSnapshot();
});
