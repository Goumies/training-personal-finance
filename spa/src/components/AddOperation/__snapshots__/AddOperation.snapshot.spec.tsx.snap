// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="add-operation"
>
  <input
    className="input-amount"
    data-e2e="account-new-operation-input-amount"
    onChange={[Function]}
    onFocus={[Function]}
    type="number"
    value=""
  />
  <button
    className="add-amount mdl-button mdl-js-button mdl-button--raised mdl-js-ripple-effect mdl-button--accent"
    data-e2e="account-new-operation-add-amount"
    onClick={[Function]}
  >
    Add
  </button>
</div>
`;
